File: 4.8-bc/go/stage-2-feature-first/app/catalog/handler.go
package catalog

import (
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-2-feature-first/shared/httputil"
)

type Handler struct {
	svc *Service
}

func NewHandler(svc *Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Get("/products", h.list)
}

func (h *Handler) list(w http.ResponseWriter, r *http.Request) {
	products, err := h.svc.List(r.Context())
	if err != nil {
		httputil.WriteError(w, err)
		return
	}
	out := make([]map[string]any, 0, len(products))
	for _, p := range products {
		out = append(out, map[string]any{
			"id":          p.ID.String(),
			"name":        p.Name,
			"price_cents": p.PriceCents,
			"category_id": p.CategoryID.String(),
		})
	}
	httputil.WriteJSON(w, http.StatusOK, map[string]any{"products": out})
}

File: 4.8-bc/py/stage-3-hexagonal/app/catalog/infra/http/handler.py
"""Catalog HTTP handler — GET /products."""
from __future__ import annotations

from fastapi import APIRouter
from pydantic import BaseModel

from app.catalog.app.service import Service


class ProductDTO(BaseModel):
    id: str
    name: str
    price_cents: int
    category_id: str


class ProductsResp(BaseModel):
    products: list[ProductDTO]


def build_router(svc: Service) -> APIRouter:
    router = APIRouter(tags=["catalog"])

    @router.get("/products", response_model=ProductsResp)
    async def list_products() -> ProductsResp:
        products = await svc.list()
        return ProductsResp(
            products=[
                ProductDTO(
                    id=str(p.id),
                    name=p.name,
                    price_cents=p.price_cents,
                    category_id=str(p.category_id),
                )
                for p in products
            ]
        )

    return router
